External IdP support - part 1

abbra opened the pull request:
I am working on a workshop chapter to add more practical details but decided to submit the PR anyway as it will take few rounds to do the review process, I guess.

There are design documents provided, they will be built automatically by ReadTheDocs bot and can be reviewed there.

Only login to console/SSH/PAM apps is implemented right now. Web UI login will not be implemented at this stage. Obtain Kerberos ticket and login with GSSAPI normally at the moment.

Todo:
 * [x] update SELinux policy according to suggestions from SELinux team
 * [x] ensure authselect upstream releases  `with-gssapi` feature support to Fedora and RHEL
 * [x] update spec file to add dependency on SSSD 2.7.0 and new authselect version
 * [x] update PR CI images to include SSSD 2.7.0 and fixed MIT Kerberos from Fedora
 * [x] add workshop chapter with practical use of External IdP feature and troubleshooting

abbra commented:
Documentation is available here: https://freeipa--6248.org.readthedocs.build/en/6248/

sumit-bose commented on ipaserver/plugins/baseuser.py:
```diff
@@ -376,6 +387,14 @@ class baseuser(LDAPObject):
+            ),
+        Str('ipaidpsub?',
+            cli_name='idp_user',
+            label=_('External IdP user name'),
```
Maybe it would be better to call it more generic 'External IdP user identifier' and the option 'idp_user_id', since even with the templates it is typically the email and not the name. It might be even worth to make this better with the option of idp-{add|mod} and replace '--subject' with '--user-id-attribute' so that 'user-id' is used in both places.

abbra replied:
Thanks. I changed it to `--idp-user-id` and updated doc/label strings to get it better represented.

sumit-bose commented on ipaserver/plugins/idp.py:
```diff
@@ -0,0 +1,442 @@
+        Str('ipaidporg?',
+            cli_name='organization',
+            label=_('Organization'),
+            doc=_('Organization ID or Realm name for IdP provider templates'),
```
Ok, my original thinking was that 'Organization' and 'Realm' more sound like the Azure domain name should be added. But now I realized that the domain name is working as well.

f-trivino commented on doc/workshop/12-external-idp-support.rst:
```diff
@@ -0,0 +1,540 @@
+protocol or console login script is not possible.
+
+OAuth 2.0 Device Authorization Grant is defined in [RFC
+8628](https://www.rfc-editor.org/rfc/rfc8628) and allows devices that either
```
[RFC
8628](https://www.rfc-editor.org/rfc/rfc8628)

perhaps:

`RFC
8628 <https://www.rfc-editor.org/rfc/rfc8628>`_



https://freeipa--6248.org.readthedocs.build/en/6248/workshop/12-external-idp-support.html

f-trivino commented on doc/workshop/12-external-idp-support.rst:
```diff
@@ -0,0 +1,540 @@
+
+In this section we set up `Keycloak <https://www.keycloak.org>`_ IdP on IPA
+client and use it to authenticate IPA users. User database in Keycloak would be
+different from the one in IPA, one would need to keey user accounts duplicated
```
keey --> keep?

f-trivino commented on doc/workshop/12-external-idp-support.rst:
```diff
@@ -0,0 +1,540 @@
+
+First, we would download keycloak and unpack it into ``/opt/keycloak-<VERSION>`` as ``root``::
+
+  dnf -y install java-11-openjdk-headless openssl
```
# dnf -y...

f-trivino commented on doc/workshop/12-external-idp-support.rst:
```diff
@@ -0,0 +1,540 @@
+
+Now we can use it for setting up users and OAuth 2.0 clients. There are two
+handy scripts, ``kcadm.sh`` and ``kcreg.sh`` that allow to perform all
+operations without visiting Web UI.
```
Keycloak Web UI

f-trivino commented on doc/workshop/12-external-idp-support.rst:
```diff
@@ -0,0 +1,540 @@
+to specify for public OAuth 2.0 clients, but they aren't used for OAuth 2.0
+device authorization grant flow.
+
+Two attributes specified in the OAuth 2.0 client definition for Keycloak enable
```
Two attributes specified in the OAuth 2.0 client definition for Keycloak enable
OAuth 2.) device authorization grant processing and set the polling interval for
the client


Notice the two attributes specified in the OAuth 2.0 client definition for Keycloak: enable
oauth2.device.authorization.grant.enabled and set oauth2.device.polling.interval to 5 as the polling interval for
the client

f-trivino commented on doc/workshop/12-external-idp-support.rst:
```diff
@@ -0,0 +1,540 @@
+OAuth 2.) device authorization grant processing and set the polling interval for
+the client. Keycloak 17.0.0 and 18.0.0 releases have a bug that sets default
+polling interval to 600 seconds. This makes impossible actual polling process as
+the lifespan of the device code is also set to 600 seconds. Keycloak's [pull
```
`pull request 11893 <https://github.com/keycloak/keycloak/pull/11893>'_

f-trivino commented on doc/workshop/11-kerberos-ticket-policy.rst:
```diff
@@ -254,12 +254,70 @@ For example, an application might decide to only allow access to a specialized
+
+Once this change made and SSSD configuration updated to allow PAM services to
+use GSSAPI authentication, it will be possible to use Kerberos ticket to
+authenticate over a choosen PAM service. A session below demonstrates it::
```
chosen

f-trivino commented on client/man/default.conf.5:
```diff
@@ -160,6 +160,9 @@ Specifies the mode the server is running in. The currently support values are \f
 Specifies the mount point that the development server will register. The default is /ipa/
 .TP
+.B oidc_child_debug_level <debuglevel>
+Specifies the debug level of \fBoidc_child\fR, a helper process used by \fBipa-otpd\fR for OIDC/OAuth2 authentication. Level can be between 0 and 10, the higher the more details. If the level is 6 or higher HTTP debug outout is added as well.
```
outout --> output

f-trivino reviewed (commented):
LGTM.

abbra commented:
@f-trivino thanks! Hopefully, I fixed all the notes. I also added troubleshooting section. There is not much to go into details there so I only showed the structure of the log output and how to crank up the debug level to see libcurl's output as well.

abbra commented:
I think this update has all changes now.
I chose to not update authselect dependency. authselect 1.4 is only released to Rawhide right now, RHEL will get 1.2.5 or 1.4 depending on the version. We don't have exact dependency to the new features yet. Once new authselect is in all Fedora and RHEL releases, we can have a separate PR that adds `with-subid` and `with-gssapi` support to IPA client installer.

f-trivino commented:
@abbra I've tested this PR on my end and worksforme. LGTM.

abbra commented:
master:

* 79a4073730a8fe5ba2424f3896a2fd440c17ac9e docs: tune RTD to display lists with disc and left margin
* 0484949b80d477ad858274b799ea1f48f2eec20a doc/designs: add External IdP support design documents
* fd19bdfd54e674361b9dadd3792780406c8d82d6 external-idp: add LDAP schema, indices and other LDAP objects
* 10e18c3dc732a52d173e803970f6eb53dd9b6087 external-idp: add support to manage external IdP objects
* 03a905eed92083b1edea634ce50fcc9dbeb34b5b external-idp: add XMLRPC tests for External IdP objects and idp indicator
* 3f6656e09a528b3f54281d77a6226231ac1c0f51 ipa-otpd: add support for SSSD OIDC helper
* a1be4fc86390559e6464fb31b76a4595da9f5465 KDB: support external IdP configuration
* 94f7d31d2dc725ebcb5a6859d32d602935c1b3b3 External IdP: add Web UI to manage IdP references
* 429e523de675f86accd8667287cc468c8f9d1872 External IdP: initial SELinux policy
* 82175da4b1c91516495a4f38a46a08ccfca4cd75 doc/workshop: document use of pam_sss_gss PAM module
* 543040a71d09710f817b29076cb7aa86d3014a02 freeipa.spec.in: use SSSD 2.7.0 to add IdP pre-auth mechanism
* b5be7f2948f72b93fc418eb2697fe96efe14a11f workshop: add chapter 12: External IdP support
